### PR TITLE
MOD-15565 -  mrange aggregation execution in inner shards

### PR DESCRIPTION
**Pull request by TalBarYakar**
> [!NOTE]
> **High Risk**
> Changes cluster MRANGE merge semantics, LibMR wire format, and multi-agg parsing—incorrect stride or coordinator re-agg would silently wrong results across shards.
> 
> **Overview**
> Cluster **TS.MRANGE** / **TS.MREVRANGE** now push **AGGREGATION**, **FILTERBY**, and alignment to each shard via LibMR; shards run the same range logic as standalone, and the coordinator merges pre-aggregated series instead of re-aggregating raw data.
> 
> The coordinator path uses **`RangeArgs.skipAggregation`** and **`RangeArgsSkipReAggregation`** so replies treat shard output as final buckets (full time window, no duplicate FILTERBY). **`SeriesListRecord.numAggClasses`** and updated **`SeriesListReplyParser`** / **`ParseSeriesAllAggs`** keep a fixed per-key stride for multi-aggregator replies and empty keys in range. **`ReplyMultiAggSeriesGroup`** emits combined `[ts, v1, v2, …]` samples; GROUPBY still reduces on the coordinator over per-series pre-aggregates (avg-of-avgs semantics). **`QueryPredicates_Arg`** serialization carries the new shard fields; CI pack/S3 steps run for the **`gen`** topology shard when topologies are split.
> 
> Extensive flow tests cover shard agg types, GROUPBY, multi-agg MRANGE, EMPTY gaps, and stride regressions.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 225143f14e065e7916b40587e0de3b2c20f4f4d2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by gabsow** [comment]
Update the PR title and use ticket number

**Comment by gabsow** [comment]
Add pr descation

File: src/libmr_integration.c [hunk 1]
```diff
@@ -788,6 +838,24 @@ static void TS_INTERNAL_MRANGE(RedisModuleCtx *ctx, void *args) {
...
 
+    // Level 1+2: apply aggregation on shard regardless of GROUPBY.
+    // The coordinator will receive pre-aggregated samples and skip re-aggregation.
+    AggregationClass *aggClasses[TS_AGG_TYPES_MAX] = { 0 };
+    if (queryArg->numAggClasses > 0) {
+        for (size_t i = 0; i < queryArg->numAggClasses; i++) {
+            aggClasses[i] = GetAggClass(queryArg->aggTypes[i]);
+        }
+        mrangeArgs.rangeArgs.aggregationArgs.numClasses = queryArg->numAggClasses;
+        mrangeArgs.rangeArgs.aggregationArgs.classes = aggClasses;
```
Comment (gabsow) [inline src/libmr_integration.c:856]: **Multi-aggregator MRANGE will crash the cluster.** When `numAggClasses > 1` (e.g. `TS.MRANGE - + AGGREGATION avg,sum 10 FILTER k=v` — allowed without GROUPBY), each shard reply samples are produced by `ReplyWithMultiAggSample` as `[ts, v1, v2, ...]`, i.e. `1 + num_values` elements (`reply.c:148-152`, `reply.c:232`).

But the coordinator's `ParseSeries` asserts every sample has exactly 2 elements:
```c
// src/libmr_integration.c:935
RedisModule_Assert(sampleElement->elements == 2);
```
so the process aborts. Even without the assert, the intermediate `Series` only stores one value per timestamp (`SeriesAddSample(result, timestamp, value)`), so multi-value buckets cannot be carried across the coordinator at all.

This is exercised by the new `test_multi_agg_avg_sum` / `test_multi_agg_min_max_count` / `test_multi_agg_mrevrange` tests — they will crash when run against a cluster. Multi-aggregator queries need to either be excluded from shard pre-aggregation (fall back to raw shipping) or `ParseSeries`/the reply parser must be taught the multi-value sample shape.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/libmr_integration.c [hunk 2]
```diff
@@ -198,6 +198,29 @@ static void QueryPredicates_ArgSerialize(WriteSerializationCtx *sctx, void *arg,
             SerializationCtxWriteRedisString(sctx, predicate->valuesList[value_index], error);
         }
     }
+
+    // per-shard aggregation fields
+    MR_SerializationCtxWriteLongLong(sctx, predicate_list->numAggClasses, error);
```
Comment (gabsow) [inline src/libmr_integration.c:203]: **These new fields are now serialized for `TS.MGET` and `TS.QUERYINDEX` too, where they are uninitialized garbage.** `QueryPredicates_ArgSerialize` is the single serializer registered for *all* of `QueryPredicatesType` — MRANGE, MGET and QUERYINDEX (`libmr_integration.c:1221-1223`). But only `TSDB_mrange_MR` initializes the new aggregation fields; `TSDB_mget_MR` (`libmr_commands.c:533`) and `TSDB_queryindex_MR` (`libmr_commands.c:669`) allocate `queryArg` with `malloc` and never touch `numAggClasses`/`aggTypes`/`filterBy*`.

So for every cluster `TS.MGET`/`TS.QUERYINDEX` on the internal protocol, this writes an uninitialized `numAggClasses`, and the loop right below reads `aggTypes[i]` for `i < numAggClasses` — with garbage `numAggClasses` that is an out-of-bounds read and can emit a multi-GB payload or segfault (likewise the garbage `filterByTSArgs.hasValue`/`count`). On the receiving shard `GetAggClass(garbage)` may return NULL and aggregation gets applied to a plain MGET.

Fix: initialize the new fields (set `numAggClasses = 0`, `filterByValueArgs.hasValue = false`, `filterByTSArgs.hasValue = false`) in `TSDB_mget_MR` and `TSDB_queryindex_MR` — or switch those allocations to `calloc`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/libmr_integration.c [hunk 3]
```diff
@@ -303,6 +326,33 @@ static void *QueryPredicates_ArgDeserialize_impl(ReaderSerializationCtx *sctx,
         goto err;
     }
 
+    // per-shard aggregation fields
+    predicates->numAggClasses = MR_SerializationCtxReadLongLong(sctx, error);
+    for (size_t i = 0; i < predicates->numAggClasses; i++) {
```
Comment (gabsow) [inline src/libmr_integration.c:333]: **Wire-controlled counts are written into fixed-size arrays with no bounds check.** `numAggClasses` is read from the stream and used to index `aggTypes[TS_AGG_TYPES_MAX]`; likewise `filterByTSArgs.count` (line 346) indexes `values[MAX_TS_VALUES_FILTER]` (128). Neither is bounded against its array size, and unlike every other read in this function the `*error` check is placed *after* the loops (line 352) rather than inside, so a corrupt/truncated/version-mismatched payload (or the garbage payload from the uninitialized MGET/QUERYINDEX path above) drives the loop to completion and overflows the embedded array → memory corruption on the shard.

Suggest bounding both before the loops, e.g.:
```c
predicates->numAggClasses = MR_SerializationCtxReadLongLong(sctx, error);
if (unlikely(*error || predicates->numAggClasses > TS_AGG_TYPES_MAX)) {
    goto err;
}
```
and similarly `if (... || predicates->filterByTSArgs.count > MAX_TS_VALUES_FILTER) goto err;` before the `values[]` loop.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/libmr_integration.c [hunk 4]
```diff
@@ -788,6 +838,24 @@ static void TS_INTERNAL_MRANGE(RedisModuleCtx *ctx, void *args) {
...
+    AggregationClass *aggClasses[TS_AGG_TYPES_MAX] = { 0 };
+    if (queryArg->numAggClasses > 0) {
+        for (size_t i = 0; i < queryArg->numAggClasses; i++) {
+            aggClasses[i] = GetAggClass(queryArg->aggTypes[i]);
+        }
+        mrangeArgs.rangeArgs.aggregationArgs.numClasses = queryArg->numAggClasses;
+        mrangeArgs.rangeArgs.aggregationArgs.classes = aggClasses;
+        mrangeArgs.rangeArgs.aggregationArgs.timeDelta = queryArg->aggTimeDelta;
+        mrangeArgs.rangeArgs.aggregationArgs.bucketTS = queryArg->aggBucketTS;
+        mrangeArgs.rangeArgs.aggregationArgs.empty = queryArg->aggEmpty;
```
Comment (gabsow) [inline src/libmr_integration.c:859]: **Verify `EMPTY` + `GROUPBY` across shards.** The coordinator reduce (`ResultSet_ApplyReducer` → `MultiSeriesAggDupSampleIterator`) merges pre-aggregated series by matching timestamps. This is only equivalent to the non-cluster path if every series in a group emits buckets at the same boundaries. With `EMPTY`, each series fills empty buckets per-shard; if a series' effective range is clamped (e.g. `RETENTION` trimming `startTimestamp`), its leading empty/NaN buckets may be absent while another group member's are present, so the timestamp-keyed merge can produce a different leading-bucket set/values than master (which aggregated all group members under one coordinator-side iterator). Worth a flow test: two same-group series on different shards, one with `RETENTION`, `AGGREGATION sum 10 EMPTY ... GROUPBY ... REDUCE sum`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/libmr_commands.c [hunk 5]
```diff
@@ -181,6 +181,17 @@ static void mrange_done_internal(ExecutionCtx *eCtx, RedisModuleCtx *ctx, MRange
         ReplyWithMapOrArray(ctx, totalLen, false);
     }
 
+    // Level 1+2: shard pre-aggregated; strip agg args so coordinator doesn't re-aggregate.
+    // RESP3 `aggregators` metadata will show empty for this path; fix when RESP3+cluster matters.
```
Comment (gabsow) [inline src/libmr_commands.c]: The "fix when RESP3+cluster matters" caveat is already live, not hypothetical: `resp3` is captured (`queryArg->resp3 = _ReplyMap(ctx)`) and this whole path only runs in a cluster. Since `coordArgs.aggregationArgs.numClasses` is set to 0 here, `ReplySeriesArrayPos`/`reply.c:117` emits `aggregators => []` for RESP3 clients, whereas standalone returns the real aggregator names (e.g. `["avg"]`). So any RESP3 client doing `TS.MRANGE ... AGGREGATION ...` against a cluster silently gets empty aggregator metadata — a client-visible divergence between cluster and standalone, not just a deferred nicety. Worth at least carrying the aggregator names through for the reply even though the values are pre-aggregated.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/libmr_integration.h [hunk 6]
```diff
@@ -30,6 +31,16 @@ typedef struct QueryPredicates_Arg
     RedisModuleString **limitLabels;
     bool latest;
     bool resp3;
+    // Per-shard aggregation fields — applied on shards for both GROUPBY and non-GROUPBY paths
```
Comment (gabsow) [inline src/libmr_integration.h:34]: _Altitude:_ these 9 fields flatten data that already lives as a coherent unit on `RangeArgs` (`aggregationArgs`, `alignment`, `timestampAlignment`, `filterByValueArgs`, `filterByTSArgs`). The cost is that the same field set must now be kept in lock-step across four hand-written sites — this struct, the pack in `TSDB_mrange_MR`, the serialize, and the deserialize — with no compiler check tying them together (the MGET/QUERYINDEX init bug above is a direct symptom). Embedding the relevant sub-structs (everything except the un-serializable `AggregationClass **classes`, which stays as the separate `aggTypes[]`) would collapse the pack/unpack to struct assignments and shrink the serde to one helper pair, removing the drift risk.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/libmr_commands.c [hunk 7]
```diff
@@ -191,15 +202,15 @@ static void mrange_done_internal(ExecutionCtx *eCtx, RedisModuleCtx *ctx, MRange
...
                                     args->reverse,
                                     false);
         });
     });
 
     if (args->groupByLabel) {
-        // Apply the reducer
-        RangeArgs rangeArgs = args->rangeArgs;
+        // Apply the reducer on pre-aggregated data (agg already stripped in coordArgs)
+        RangeArgs rangeArgs = coordArgs;
```
Comment (gabsow) [inline src/libmr_commands.c:236]: _Cleanup:_ just below, `minimizedArgs` (line 218) re-derives from `args->rangeArgs` and re-zeroes the same agg/filter fields that `coordArgs` (lines 186-192) already stripped. It could start from the already-stripped copy and only set the extra overrides:
```c
RangeArgs minimizedArgs = coordArgs;
minimizedArgs.startTimestamp = 0;
minimizedArgs.endTimestamp = UINT64_MAX;
minimizedArgs.aggregationArgs.timeDelta = 0;
minimizedArgs.latest = false;
```
There's now a third near-identical "strip pre-aggregated args" block in `module.c:388` (already tagged `// todo: this is duplicated`); a shared helper would keep the field set from drifting across the three sites.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by gabsow — COMMENTED** [review]
Code review — cluster MRANGE shard pre-aggregation (high-effort, recall-biased).

The core design is sound: a series lives entirely on one shard, so per-series aggregation completes shard-side, the coordinator correctly stops re-aggregating, and the GROUPBY reduce-over-pre-aggregated-buckets is semantically equivalent to the standalone per-series-then-reduce path (I verified the non-cluster path also produces avg-of-per-series-avgs, so the documented "Option A" is *not* a divergence). COUNT, MREVRANGE ordering, and FILTER-without-AGGREGATION all check out.

Two blocking correctness issues found:

1. **Multi-aggregator MRANGE crashes the cluster** — shards emit `[ts,v1,v2]` multi-value samples but `ParseSeries` asserts exactly 2 elements (and the coordinator `Series` can't hold multi-value buckets). The new `test_multi_agg_*` tests will abort against a cluster.
2. **`TS.MGET`/`TS.QUERYINDEX` serialize uninitialized garbage** — the new fields are added to the shared serializer but only initialized on the MRANGE path; MGET/QUERYINDEX `malloc` without zeroing → OOB read / corruption on every cluster MGET/QUERYINDEX.

Plus a deserialization bounds-check gap, an EMPTY+GROUPBY+RETENTION alignment edge to verify, a live RESP3 `aggregators`-metadata regression, and two cleanup/altitude notes. Details inline.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/libmr_commands.c [hunk 8]
```diff
@@ -603,6 +614,17 @@ int TSDB_mrange_MR(RedisModuleCtx *ctx, RedisModuleString **argv, int argc, bool
     }
 
     queryArg->userName = CopyCurrentUserName(ctx);
+    queryArg->numAggClasses = args.rangeArgs.aggregationArgs.numClasses;
```
[the remark this answers is not in the document]
Comment (TalBarYakar) [reply src/libmr_commands.c]: done

File: src/libmr_integration.c [hunk 9]
```diff
@@ -303,6 +326,33 @@ static void *QueryPredicates_ArgDeserialize_impl(ReaderSerializationCtx *sctx,
...
+    if (predicates->filterByValueArgs.hasValue) {
+        predicates->filterByValueArgs.min = MR_SerializationCtxReadDouble(sctx, error);
+        predicates->filterByValueArgs.max = MR_SerializationCtxReadDouble(sctx, error);
+    }
+    predicates->filterByTSArgs.hasValue = MR_SerializationCtxReadLongLong(sctx, error);
+    if (predicates->filterByTSArgs.hasValue) {
+        predicates->filterByTSArgs.count = MR_SerializationCtxReadLongLong(sctx, error);
+        for (size_t i = 0; i < predicates->filterByTSArgs.count; i++) {
+            predicates->filterByTSArgs.values[i] = MR_SerializationCtxReadLongLong(sctx, error);
+        }
```
[the remark this answers is not in the document]
Comment (TalBarYakar) [reply src/libmr_integration.c:357]: Both reads now cap at their array bounds before the loop — any oversized value from a malformed payload hits goto err instead of writing past the array.

File: src/libmr_integration.c (hunk 1 above)
Comment (TalBarYakar) [reply src/libmr_integration.c:856]: ParseSeries is replaced by ParseSeriesAllAggs which handles the [ts, v0, v1, ...] shape. The function detects numAgg from the first sample's element count (first->elements - 1) — no hardcoded == 2 assertion. For N agg types it allocates N Series per key: series[0] carries name + labels and stores v0 values, series[1..N-1] carry name only and store v1..vN-1. Timestamps are shared.

SeriesListRecord gains a numAggClasses field so the coordinator knows the per-key stride. On the reply side, mrange_done_internal groups seriesList in strides of numAggTypes and routes multi-agg keys to a new ReplyMultiAggSeriesGroup function, which opens N lockstep sample iterators and emits [ts, v0, v1, ..., vN-1] directly — no intermediate heap allocation.

Single-agg and no-agg keys (numAggTypes == 1) continue through the existing ReplySeriesArrayPos path unchanged.

File: src/libmr_integration.c (hunk 2 above)
Comment (TalBarYakar) [reply src/libmr_integration.c:203]: Fixed. TSDB_mget_MR and TSDB_queryindex_MR now use calloc instead of malloc — all fields including numAggClasses, filterByValueArgs.hasValue, and filterByTSArgs.hasValue are zeroed, so the serializer writes numAggClasses=0 and skips the agg/filter loops for those commands.

File: src/libmr_integration.c (hunk 3 above)
Comment (TalBarYakar) [reply src/libmr_integration.c:333]: fixed

File: src/libmr_integration.c (hunk 4 above)
Comment (TalBarYakar) [reply src/libmr_integration.c:859]: EMPTY only fills gaps between existing samples and the trailing suffix after the last sample; it does not synthesize buckets before the series' first sample. This is anchored in agg_iter_apply_empty_prefix (filter_iterator.c) which computes first_bucket from startTimestamp but only emits empty buckets up to CalcBucketStart(first_sample_ts) — if the series has no samples at all in a leading range, first_sample_ts is whatever the first retained sample is and nothing is emitted before it.

So for a series with RETENTION that trims data before t=70, querying MRANGE 0 100 AGGREGATION sum 10 EMPTY produces [70, 80, 90] — not [0, 10, ..., 60, 70, 80, 90]. The standalone path sees the same thing (RETENTION already deleted those samples on the shard before the coordinator ever touches them), so the coordinator also never aggregates [0-69] for that series. Both paths are equivalent. No divergence.

File: src/libmr_commands.c (hunk 5 above)
Comment (TalBarYakar) [reply src/libmr_commands.c]: fixed.
added a const AggregationArgs *displayAggs parameter to ReplyMultiAggSeriesGroup. The caller now passes:

replyArgs (stripped) → used for the actual data query (no re-aggregation)
&args->rangeArgs.aggregationArgs (original) → used only for the RESP3 aggregators display
The two concerns are now decoupled without touching shared structs or any other function signatures.

File: src/libmr_integration.h (hunk 6 above)
Comment (TalBarYakar) [reply src/libmr_integration.h:34]: Good point, we'll do this in a follow-up PR to keep this one focused. The change touches more spots than expected (serde, macro, mget unpack on top of the struct def and pack site).

File: src/libmr_commands.c (hunk 7 above)
Comment (TalBarYakar) [reply src/libmr_commands.c:236]: Fixed both. minimizedArgs now starts from coordArgs and calls RangeArgs_ZeroProcessing() which is shared with the module.c site — no more drift.

File: src/libmr_commands.c [hunk 10]
```diff
@@ -181,6 +181,17 @@ static void mrange_done_internal(ExecutionCtx *eCtx, RedisModuleCtx *ctx, MRange
...
+    // Level 1+2: shard pre-aggregated; strip agg args so coordinator doesn't re-aggregate.
+    // RESP3 `aggregators` metadata will show empty for this path; fix when RESP3+cluster matters.
+    RangeArgs coordArgs = args->rangeArgs;
+    if (coordArgs.aggregationArgs.numClasses > 0) {
+        coordArgs.aggregationArgs.numClasses = 0;
+        coordArgs.aggregationArgs.classes = NULL;
+        coordArgs.filterByValueArgs.hasValue = false;
+        coordArgs.filterByTSArgs.hasValue = false;
+    }
+    const RangeArgs *replyArgs = &coordArgs;
```
[the remark this answers is not in the document]
Comment (TalBarYakar) [reply src/libmr_commands.c:201]: fixed

File: src/libmr_integration.c [hunk 11]
```diff
@@ -198,6 +198,29 @@ static void QueryPredicates_ArgSerialize(WriteSerializationCtx *sctx, void *arg,
...
+        MR_SerializationCtxWriteDouble(sctx, predicate_list->filterByValueArgs.min, error);
+        MR_SerializationCtxWriteDouble(sctx, predicate_list->filterByValueArgs.max, error);
+    }
+    MR_SerializationCtxWriteLongLong(sctx, predicate_list->filterByTSArgs.hasValue, error);
+    if (predicate_list->filterByTSArgs.hasValue) {
+        MR_SerializationCtxWriteLongLong(sctx, predicate_list->filterByTSArgs.count, error);
+        for (size_t i = 0; i < predicate_list->filterByTSArgs.count; i++) {
+            MR_SerializationCtxWriteLongLong(sctx, predicate_list->filterByTSArgs.values[i], error);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (TalBarYakar) [reply src/libmr_integration.c:223]: fixed

File: src/libmr_commands.c [hunk 12]
```diff
@@ -167,52 +166,75 @@ static void mrange_done_internal(ExecutionCtx *eCtx, RedisModuleCtx *ctx, MRange
...
+        coordArgs.skipAggregation = true;
+        coordArgs.filterByValueArgs.hasValue = false;
+        coordArgs.filterByTSArgs.hasValue = false;
+    }
+    const RangeArgs *replyArgs = &coordArgs;
+
+    array_foreach(nodesResults, record, {
+        ARR(Series *) sl = record->seriesList;
+        size_t numAggTypes = (preAgg && record->numAggClasses > 1) ? record->numAggClasses : 1;
+        size_t numKeys = array_len(sl) / numAggTypes;
```
Comment (gabsow) [inline src/libmr_commands.c:206]: **New bug from the multi-agg reassembly: an empty series in a multi-aggregator query drops/mis-pairs data.** This reassembly assumes every matched key contributed exactly `numAggTypes` consecutive `Series`. But `ParseSeriesAllAggs` derives `numAgg` *per series* from the first sample's shape, and a filter-matched series with **no samples in the queried range** is still emitted by the shard as `[name, labels, []]` (`module.c:466` replies every key unconditionally). For that empty series `samplesElement->elements == 0`, so `ParseSeriesAllAggs` returns `numAgg = 1`, not `numAggTypes`.

That breaks the fixed-stride grouping here:
```c
size_t numAggTypes = (preAgg && record->numAggClasses > 1) ? record->numAggClasses : 1;
size_t numKeys = array_len(sl) / numAggTypes;   // floor → misaligned when any series is empty
...
Series **group = &sl[k * numAggTypes];
```

Concrete repro (cluster): `s1` has data in `[0,29]`, `s2` matches the same filter but has no samples in `[0,29]`, then
`TS.MRANGE 0 29 AGGREGATION avg,sum 10 FILTER k=v`. Shard returns `[[s1,labels,[[0,15,30]...]], [s2,labels,[]]]` → flattened `[s1_avg, s1_sum, s2]` (len 3), `numAggTypes=2`, `numKeys = 3/2 = 1`. Depending on dict order either `s2` is silently dropped, or `group = {s2, s1_avg}` mixes two different keys — `ReplyMultiAggSeriesGroup` then emits `s2`'s name/labels with `s1`'s iterator and loses `s1_sum`.

The existing `test_multi_agg_*` tests don't catch this because every series has data in range. Fix: don't infer per-series `numAgg` from sample shape — carry the authoritative `numAggClasses` from the request and have `ParseSeriesAllAggs` always materialize that many `Series` (empty ones included), or group by key name rather than a fixed stride. Worth a regression test with a multi-agg filter that matches an out-of-range/empty series.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by gabsow — COMMENTED** [review]
Re-review of the latest push (`3390d6b`). The earlier findings are well addressed:

- **Multi-agg crash** → resolved with `ParseSeriesAllAggs` + `ReplyMultiAggSeriesGroup`.
- **MGET/QUERYINDEX uninitialized agg fields** → resolved (`calloc`).
- **Unbounded deserialization** → resolved (`numAggClasses`/`filterByTSArgs.count` bounds → `goto err`).
- **RESP3 `aggregators` empty** → resolved via `skipAggregation` (keeps agg metadata, skips re-agg).
- **Duplicated agg-strip** → resolved with `RangeArgs_ZeroProcessing`.

One **new correctness bug** in the multi-agg reassembly (empty/out-of-range series breaks the fixed-stride grouping) — details inline. Also still open: the EMPTY + GROUPBY + RETENTION bucket-alignment edge isn't covered by the new EMPTY tests (non-GROUPBY only). I'd hold merge on the inline finding.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/libmr_commands.c [hunk 13]
```diff
@@ -163,56 +162,85 @@ static void *collect_node_results(ExecutionCtx *eCtx, RedisModuleCtx *ctx) {
...
-            else
+    // In the pre-agg path the coordinator must not re-aggregate or re-filter shard data.
+    // Keep agg info intact so RESP3 can display the aggregator names.
+    RangeArgs coordArgs = preAgg ? RangeArgsSkipReAggregation(&args->rangeArgs) : args->rangeArgs;
+    const RangeArgs *replyArgs = &coordArgs;
+
+    array_foreach(nodesResults, record, {
+        ARR(Series *) sl = record->seriesList;
+        size_t numAggTypes = (preAgg && record->numAggClasses > 1) ? record->numAggClasses : 1;
+        size_t numKeys = array_len(sl) / numAggTypes;
```
Comment (gabsow) [inline src/libmr_commands.c:206]: Re-flagging (my earlier comment is now outdated by the line shift — this is **still unfixed** in `503ad21`): the fixed-stride regrouping `numKeys = array_len(sl) / numAggTypes` + `group = &sl[k * numAggTypes]` assumes every matched key contributed exactly `numAggTypes` series. But `ParseSeriesAllAggs` still infers `numAgg` per-series from the first sample's shape (libmr_integration.c:932-937), so a filter-matched series with **no samples in the queried range** — which `TS.INTERNAL_MRANGE_AGG` → `replyUngroupedMultiRange` still emits as `[name, labels, []]` — parses as `numAgg=1`, not `numAggTypes`. The stride then misaligns and a series is dropped or two different keys get merged into one multi-agg reply.

Repro (cluster): `s1` has data in `[0,29]`, `s2` shares the filter but has samples only outside `[0,29]`, then `TS.MRANGE 0 29 AGGREGATION avg,sum 10 FILTER k=v`. The 4 `test_multi_agg_*` tests don't catch it because every series has in-range data.

Fix: have `ParseSeriesAllAggs` always materialize the authoritative `numAggClasses` series (carried in `QueryPredicates_Arg`) including empty ones, rather than inferring from sample shape — or regroup by key name instead of a fixed stride. Please add a regression test with a multi-agg filter matching an empty/out-of-range series.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by gabsow — COMMENTED** [review]
Re-review of `503ad21`. The new `TS.INTERNAL_MRANGE_AGG` command (so old shards return a clear "unknown command" in mixed-version clusters) is a good addition and addresses the wire-format/rollout concern.

However, the **multi-agg empty-series bug from my previous review is still unfixed** — the reassembly stride and `ParseSeriesAllAggs`'s shape-based `numAgg` inference are unchanged, and there's still no regression test for a multi-agg query matching an empty/out-of-range series. Details inline. I'd hold merge on this one.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by gabsow** [comment]
### Review status — `503ad21`

Most of the earlier findings are resolved. Nice work on the rollout safety in particular.

**Resolved**
- Multi-agg samples no longer crash the coordinator (`ParseSeriesAllAggs` + `ReplyMultiAggSeriesGroup`).
- `TS.MGET`/`TS.QUERYINDEX` no longer serialize uninitialized agg fields (`calloc`).
- Deserializer bounds-checks `numAggClasses` / `filterByTSArgs.count` before indexing the fixed arrays.
- RESP3 `aggregators` metadata preserved on the cluster path (`skipAggregation` / `RangeArgsSkipReAggregation`).
- Agg-strip duplication factored into `RangeArgs_ZeroProcessing`.
- New `TS.INTERNAL_MRANGE_AGG` command gives mixed-version clusters a clean "unknown command" error instead of silent misbehavior. 👍

**Still blocking (1)**
- **Multi-agg + empty/out-of-range series produces wrong results.** The coordinator regroups shard replies by a fixed stride (`array_len(sl) / numAggTypes`), but `ParseSeriesAllAggs` infers the per-series agg count from the first sample's shape — so a filter-matched series with no samples in the queried range (still emitted by the shard as `[name, labels, []]`) parses as 1 series instead of `numAggClasses`, misaligning the stride. A series gets dropped or two keys get merged into one reply. Repro and suggested fix are in the inline comment on `src/libmr_commands.c`. The existing `test_multi_agg_*` tests don't catch it because every series has in-range data — please add a regression test for the empty/out-of-range case.

**Open, low priority (not a confirmed bug)**
- EMPTY + GROUPBY + RETENTION bucket-alignment across shards isn't covered by the new EMPTY tests (those are non-GROUPBY). Worth a targeted test to confirm parity with standalone.

I'd hold merge until the multi-agg empty-series case is fixed and tested.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/libmr_commands.c [hunk 14]
```diff
@@ -167,52 +166,75 @@ static void mrange_done_internal(ExecutionCtx *eCtx, RedisModuleCtx *ctx, MRange
...
+        coordArgs.filterByTSArgs.hasValue = false;
+    }
+    const RangeArgs *replyArgs = &coordArgs;
+
+    array_foreach(nodesResults, record, {
+        ARR(Series *) sl = record->seriesList;
+        size_t numAggTypes = (preAgg && record->numAggClasses > 1) ? record->numAggClasses : 1;
+        size_t numKeys = array_len(sl) / numAggTypes;
+        for (size_t k = 0; k < numKeys; k++) {
+            Series **group = &sl[k * numAggTypes];
```
[the remark this answers is not in the document]
Comment (TalBarYakar) [reply src/libmr_commands.c:208]: fixed

File: src/libmr_commands.c (hunk 12 above)
Comment (TalBarYakar) [reply src/libmr_commands.c:206]: fixed

File: src/libmr_integration.c [hunk 15]
```diff
@@ -303,6 +326,39 @@ static void *QueryPredicates_ArgDeserialize_impl(ReaderSerializationCtx *sctx,
...
+    predicates->filterByTSArgs.hasValue = MR_SerializationCtxReadLongLong(sctx, error);
+    if (predicates->filterByTSArgs.hasValue) {
+        predicates->filterByTSArgs.count = MR_SerializationCtxReadLongLong(sctx, error);
+        if (predicates->filterByTSArgs.count > MAX_TS_VALUES_FILTER) {
+            goto err;
+        }
+        for (size_t i = 0; i < predicates->filterByTSArgs.count; i++) {
+            predicates->filterByTSArgs.values[i] = MR_SerializationCtxReadLongLong(sctx, error);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (TalBarYakar) [reply src/libmr_integration.c:358]: fixed

File: nodes.conf [hunk 16]
```diff
@@ -0,0 +1,2 @@
+a9415fb7bd01569c24ed151531affae6c401046a :0@0,,tls-port=0,shard-id=b606941c7825f432aea7657db16f795fb89d1267 myself,master - 0 0 0 connected
+vars currentEpoch 0 lastVoteEpoch 0
```
[the remark this answers is not in the document]
Comment (TalBarYakar) [reply nodes.conf]: removed

**Comment by gabsow** [comment]
### Re-review — `78df603`

**Resolved** ✅ The multi-agg empty/out-of-range series bug is fixed cleanly: `SeriesListReplyParser` now does a first pass to find the true `numAgg` from any populated key, then re-parses every key with that as a floor (`ParseSeriesAllAggs(reply, minNumAgg)`), so empty keys produce a matching number of `Series` and the coordinator stride stays aligned. The all-empty-shard case degrades to `numAgg=1` but is harmless (empty series reply identically either way). Nice.

**New concern — mixed-version/rolling-upgrade safety regressed.** The previous push had a dedicated `TS.INTERNAL_MRANGE_AGG` command (+ a version-mismatch flow test) specifically so an old shard would return a clean "unknown command" error to a new coordinator. This push reverts to dispatching the single `TS.INTERNAL_MRANGE` (libmr_commands.c:663) and drops the version test. Consequence during a rolling upgrade (new coordinator → not-yet-upgraded shard):
- new coordinator serializes agg fields and dispatches `TS.INTERNAL_MRANGE`;
- the old shard's deserializer ignores the trailing bytes and runs the legacy path → returns **raw, unaggregated** samples;
- the new coordinator's `mrange_done_internal` unconditionally applies `RangeArgsSkipReAggregation` (skip re-agg, clear FILTERBY, open the time window), so it relays those raw samples **as if pre-aggregated/filtered** → silently wrong results, no error.

Was removing the `_AGG` command intentional (e.g. upgrades require a full restart, so transient mixed-version is out of scope)? If rolling upgrades are supported, please restore the distinct command name (or a version/capability guard) so a stale shard fails loudly instead of returning incorrect data.

**Minor**
- No regression test was added for the multi-agg + empty/out-of-range case that the parser fix addresses — worth one to lock the fix in.
- EMPTY + GROUPBY + RETENTION bucket-alignment is still only covered by non-GROUPBY EMPTY tests.

FILTERBY routing (shards always apply, coordinator never re-applies) looks consistent for the internal path now. Aside from the mixed-version question, this is close.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by gabsow — DISMISSED** [review]
Approving. The correctness blocker from earlier reviews — multi-agg with empty/out-of-range series — is properly fixed (two-pass `SeriesListReplyParser` + `minNumAgg` floor keeps the coordinator stride aligned). FILTERBY routing (shards apply, coordinator skips re-apply) is consistent for the internal path, and the earlier serialization/bounds/RESP3/dedup findings remain resolved.

Two non-blocking follow-ups to consider before/after merge:
- Mixed-version rolling upgrades: with the `TS.INTERNAL_MRANGE_AGG` command removed, a new coordinator talking to a not-yet-upgraded shard can silently get raw (unaggregated) samples relayed as if pre-aggregated. Fine to defer if module upgrades require a full restart; otherwise restore a distinct command name / version guard.
- Add a regression test for the multi-agg empty/out-of-range case, and one for EMPTY + GROUPBY + RETENTION.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by gabsow — APPROVED** [review]
